server/core/src/irodsReServer.cpp

```diff
@@ -337,63 +373,37 @@ int main() {
+    auto worker_conn_pool = std::make_shared<irods::connection_pool>(
+        thread_count,
+        env.rodsHost,
+        env.rodsPort,
+        env.rodsUserName,
+        env.rodsZone,
+        env.irodsConnectionPoolRefreshTime);
+
     while(!re_server_terminated) {
         try {
-            rodsEnv env{};
-            _getRodsEnv(env);
-            auto query_conn_pool = std::make_shared<irods::connection_pool>(
-                1,
-                env.rodsHost,
-                env.rodsPort,
-                env.rodsUserName,
-                env.rodsZone,
-                env.irodsConnectionPoolRefreshTime);
+            auto delay_queue_processor = make_delay_queue_query_processor(worker_conn_pool, queue, re_server_terminated);
+            auto query_conn_pool = irods::connection_pool::make_one();
             auto query_conn = query_conn_pool->get_connection();
-            const auto now = std::to_string(std::time(nullptr));
-            const auto qstr = (boost::format(
-                "SELECT RULE_EXEC_ID, \
-                        RULE_EXEC_NAME, \
-                        RULE_EXEC_REI_FILE_PATH, \
-                        RULE_EXEC_USER_NAME, \
-                        RULE_EXEC_ADDRESS, \
-                        RULE_EXEC_TIME, \
-                        RULE_EXEC_FREQUENCY, \
-                        RULE_EXEC_PRIORITY, \
-                        RULE_EXEC_LAST_EXE_TIME, \
-                        RULE_EXEC_STATUS, \
-                        RULE_EXEC_ESTIMATED_EXE_TIME, \
-                        RULE_EXEC_NOTIFICATION_ADDR \
-                WHERE RULE_EXEC_TIME <= '%s'") % now).str();
-            irods::query<rcComm_t> qobj{&static_cast<rcComm_t&>(query_conn), qstr};
-            if(qobj.size() > 0) {
-                rodsEnv env{};
-                _getRodsEnv(env);
-                auto conn_pool = std::make_shared<irods::connection_pool>(
-                    std::min<int>(thread_count, qobj.size()),
-                    env.rodsHost,
-                    env.rodsPort,
-                    env.rodsUserName,
-                    env.rodsZone,
-                    env.irodsConnectionPoolRefreshTime);
-                for(const auto& result: qobj) {
-                    const auto& rule_id = result[0];
-                    if(queue.contains_rule_id(rule_id)) {
-                        continue;
-                    }
-                    logger::delay_server::trace((boost::format("Enqueueing rule [%s]") % rule_id).str());
-                    queue.enqueue_rule(rule_id);
-                    irods::thread_pool::post(thread_pool, [conn_pool, &queue, result] {
-                        execute_rule(conn_pool, queue, result, re_server_terminated);
-                    });
+            auto future = delay_queue_processor.execute(thread_pool, static_cast<rcComm_t&>(query_conn));
+            auto errors = future.get();
+            if(errors.size() > 0) {
+                for(const auto [code, msg] : errors) {
```

Comment:
You can add a `&` to avoid copying (e.g. `for (const auto& [code, msg] : errors)`).

Reply:
Got it, thanks!